### PR TITLE
Attempt to recover from db with missing table files

### PR DESCRIPTION
**Pull request by riversand963**
There are situations when RocksDB tries to recover, but the db is in an inconsistent state due to SST files referenced in the MANIFEST being missing. In this case, previous RocksDB will just fail the recovery and return a non-ok status.
This PR enables another possibility. During recovery, RocksDB checks possible MANIFEST files, and try to recover to the most recent state without missing table file. `VersionSet::Recover()` applies version edits incrementally and "materializes" a version only when this version does not reference any missing table file. After processing the entire MANIFEST, the version created last will be the latest version.
`DBImpl::Recover()` calls `VersionSet::Recover()`. Afterwards, WAL replay will *not* be performed.
To use this capability, set `options.best_efforts_recovery = true` when opening the db. Best-efforts recovery is currently incompatible with atomic flush.

Test plan (on devserver):
```
$make check
$COMPILE_WITH_ASAN=1 make all && make check
```

### CODE REVIEW

File: db/version_set.cc [hunk 1]
```diff
@@ -4553,6 +4592,410 @@ Status VersionSet::Recover(
...
+        auto fiter = missing_files.find(file_num);
+        if (fiter != missing_files.end()) {
+          missing_files.erase(fiter);
+        }
+      }
+      for (const auto& elem : edit.GetNewFiles()) {
+        const FileMetaData& meta = elem.second;
+        const FileDescriptor& fd = meta.fd;
+        uint64_t file_num = fd.GetNumber();
+        const std::string fname = MakeTableFileName(dbname_, file_num);
```
Comment (riversand963) [inline db/version_set.cc]: Should consider path id when constructing table file name.

File: db/db_impl/db_impl_files.cc [hunk 2]
```diff
@@ -664,4 +665,49 @@ uint64_t PrecomputeMinLogNumberToKeep(
   return min_log_number_to_keep;
 }
 
+Status DBImpl::CleanupFilesAfterRecovery() {
```
[the remark this answers is not in the document]
Comment (riversand963) [reply db/db_impl/db_impl_files.cc]: Done.

File: db/db_impl/db_impl_files.cc [hunk 3]
```diff
@@ -664,4 +665,49 @@ uint64_t PrecomputeMinLogNumberToKeep(
...
+  mutex_.AssertHeld();
+  std::set<std::string> paths;
+  for (const auto& db_path : immutable_db_options_.db_paths) {
+    paths.insert(db_path.path);
+  }
```
[the remark this answers is not in the document]
Comment (riversand963) [reply db/db_impl/db_impl_files.cc]: Good catch. Checked `SanitizeOptions()`, and it should be added here. I will also need to dedup.

File: db/db_impl/db_impl_open.cc [hunk 4]
```diff
@@ -543,7 +553,7 @@ Status DBImpl::Recover(
       }
     }
 
-    if (!logs.empty()) {
+    if (!logs.empty() && !missing_table_file) {
```
[the remark this answers is not in the document]
Comment (riversand963) [reply db/db_impl/db_impl_open.cc]: No. It means whether the last version contains missing file, or put it another way, whether the db recovers *all* version edits in the MANIFEST. If not, there is a gap between MANIFEST and WAL, and we should not replay WAL. If missing_files = false, then we can replay WAL.

File: db/version_set.cc [hunk 5]
```diff
@@ -5536,6 +5997,21 @@ uint64_t VersionSet::GetTotalSstFilesSize(Version* dummy_versions) {
   return total_files_size;
 }
 
+Status VersionSet::VerifyFileMetadata(const std::string& fpath,
+                                      const FileMetaData& meta) const {
+  Status status = env_->FileExists(fpath);
+  uint64_t fsize = 0;
+  if (status.ok()) {
+    status = fs_->GetFileSize(fpath, IOOptions(), &fsize, nullptr);
```
[the remark this answers is not in the document]
Comment (riversand963) [reply db/version_set.cc]: I think it is an IOError with PathNotFound subcode. See https://github.com/facebook/rocksdb/blob/master/env/io_posix.h#L54:L56.
The `GetFileSize` does make `FileExists` unnecessary.

File: db/version_set.cc [hunk 6]
```diff
@@ -5536,6 +5997,21 @@ uint64_t VersionSet::GetTotalSstFilesSize(Version* dummy_versions) {
   return total_files_size;
 }
 
+Status VersionSet::VerifyFileMetadata(const std::string& fpath,
+                                      const FileMetaData& meta) const {
+  Status status = env_->FileExists(fpath);
```
[the remark this answers is not in the document]
Comment (riversand963) [reply db/version_set.cc]: This can be resolved with the next comment.

File: db/version_set.cc [hunk 7]
```diff
@@ -4553,6 +4592,422 @@ Status VersionSet::Recover(
...
+      }
+    }
+  }
+  return s;
+}
+
+// TODO Make this function shorter.
+// Also note that LoadTableHandlers should be called after we determine the
+// final version, thus may likely be outside this function.
+Status VersionSet::TryRecoverFromOneManifest(
```
[the remark this answers is not in the document]
Comment (riversand963) [reply db/version_set.cc]: I agree. Let me try again. The logic of handling each version edit is pretty different from what we do in regular recovery though. I expect a refactoring to combine `TryRecoverFromOneManifest` and `Recover` as a follow up. At this point, not sure if it is a good idea to modify `VersionSet::Recover` unnecessarily. 
but let me try again.

[the remark this answers is not in the document]
Comment (riversand963) [reply db/version_set.cc]: @anand1976, after checking the code, I believe we should have a `VersionEditHandler` class and subclasses to make the code simpler. Should we do it here or in another PR?

**Comment by riversand963** [comment]
Another way to test is to temporarily set the default value of options.incremental_recovery to true and make check.

File: db/version_set.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (riversand963) [reply db/version_set.cc]: @anand1976, I tried to address this comment to shorten the function. I can see a way of doing it, as in a temp branch: https://github.com/facebook/rocksdb/compare/master...riversand963:version-edit-handler.
If we pursue this path, I will add VersionEditHandler class in this PR, and use it only for `TryRecoverFromOneManifest`, but not in `Recover` and `ReadAndApply`. Afterwards, I will use this class and its subclasses to refactor `Recover` and `ReadAndApply`.
Could you let me know what you think? Thanks

[the remark this answers is not in the document]
Comment (riversand963) [reply db/version_set.cc]: Done with refactoring that can be found in recent commits. Ready for another round of review.

File: db/version_edit_handler.cc [hunk 8]
```diff
@@ -0,0 +1,591 @@
...
+          "MANIFEST record referencing unknown column family");
+    }
+    ColumnFamilyData* tmp_cfd = nullptr;
+    if (s.ok()) {
+      auto builder_iter = builders_.find(edit.column_family_);
+      assert(builder_iter != builders_.end());
+      tmp_cfd = version_set_->GetColumnFamilySet()->GetColumnFamily(
+          edit.column_family_);
+      assert(tmp_cfd != nullptr);
+      s = CreateVersion(edit, tmp_cfd, false);
```
Comment (anand1976) [inline db/version_edit_handler.cc]: Why create a new Version each time? Why not just apply to current?

File: db/version_edit_handler.cc [hunk 9]
```diff
@@ -0,0 +1,591 @@
...
+  // in builders means that user supplied that column family
+  // option AND that we encountered column family add record
+  bool in_builders = builders_.find(edit.column_family_) != builders_.end();
+  // They cannot both be true
+  assert(!(in_not_found && in_builders));
+  *cf_in_not_found = in_not_found;
+  *cf_in_builders = in_builders;
+}
+
+void VersionEditHandler::CheckIterateResult(const log::Reader& reader,
```
Comment (anand1976) [inline db/version_edit_handler.cc]: The name CheckIterateResult was a little confusing to me, as I thought it implied another round of iteration. Maybe CheckIterationResult would be clearer.

File: db/version_edit_handler.cc [hunk 10]
```diff
@@ -0,0 +1,591 @@
...
+    uint64_t file_num = fd.GetNumber();
+    const std::string fpath =
+        MakeTableFileName(cfd->ioptions()->cf_paths[0].path, file_num);
+    s = version_set_->VerifyFileMetadata(fpath, meta);
+    if (s.IsPathNotFound() || s.IsNotFound()) {
+      missing_files.insert(file_num);
+      s = Status::OK();
+    }
+  }
+  bool missing_info = !version_edit_params_.has_log_number_ ||
```
Comment (anand1976) [inline db/version_edit_handler.cc]: Wouldn't ```missing_info_ == true``` indicate a corrupt manifest entry? Maybe we should log a message atleast since its not expected.

File: db/version_edit_handler.h [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+#include "db/version_edit.h"
+#include "db/version_set.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+typedef std::unique_ptr<BaseReferencedVersionBuilder> VersionBuilderUPtr;
+
+// Not thread-safe, external synchronization is necessary if an object of
+// VersionEditHandler is shared by multiple threads.
+class VersionEditHandler {
```
Comment (anand1976) [inline db/version_edit_handler.h]: Some comments here about what the class does would be good, especially how to call it and how it interacts with VersionSet.

File: db/version_edit_handler.h [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+      cf_to_missing_files_;
+  bool ignore_missing_files_;
+
+ private:
+  Status ExtractInfoFromVersionEdit(ColumnFamilyData* cfd,
+                                    const VersionEdit& edit);
+  bool initialized_;
+};
+
+class VersionEditHandlerPointInTime : public VersionEditHandler {
```
Comment (anand1976) [inline db/version_edit_handler.h]: Comments here too

File: db/version_edit_handler.cc [hunk 13]
```diff
@@ -0,0 +1,591 @@
...
+VersionEditHandlerPointInTime::~VersionEditHandlerPointInTime() {
+  for (const auto& elem : versions_) {
+    delete elem.second;
+  }
+  versions_.clear();
+}
+
+void VersionEditHandlerPointInTime::CheckIterateResult(
+    const log::Reader& reader, Status* s) {
+  VersionEditHandler::CheckIterateResult(reader, s);
```
Comment (anand1976) [inline db/version_edit_handler.cc]: Should this be called here or after appending the versions to ```version_set_```?

File: db/version_edit_handler.cc (hunk 10 above)
Comment (riversand963) [reply db/version_edit_handler.cc]: No it does not indicate one corrupted MANIFEST entry.
If missing_info is true after processing the entire MANIFEST, then the MANIFEST file is incomplete, and we cannot recover.

File: db/version_edit_handler.cc (hunk 13 above)
Comment (riversand963) [reply db/version_edit_handler.cc]: Which function do you mean?
`VersionEditHandler::CheckIterateResult` should be called before appending the versions to version_set_ because it is necessary to perform some checks to see if the recovery is successful.

Comment (anand1976) [reply db/version_edit_handler.cc]: Ok. I got confused because ```VersionEditHandler::CreateVersion``` creates and appends a new version to ```version_set_```, so it would already have a set of valid versions when ```VersionEditHandler::CheckIterateResult``` is called.

File: db/version_edit_handler.cc (hunk 8 above)
Comment (riversand963) [reply db/version_edit_handler.cc]: We currently rely on `VersionBuilder::SaveTo(version->strorage_info())` to create versions. 
The version builder keeps a history of all added and delete files. Calling `version_builder->SaveTo(version->storage_info())` repeatedly for the same version will cause the same files be added.
This is also our current way of building versions
1. Create a version builder
2. version builder applies multiple version edits
3. Finally, create a version, call version_builder->SaveTo(version)
4. Append version
I am following this practice.

**Comment by anand1976** [comment]
@riversand963 Thanks for the explanation about the reasoning behind handling CF add/drop. With this recovery, we don't support cross CF consistency. We may have to revisit later for cases when WAL is used or atomic flush is enabled. For now, should we just not try to recover if either of those is true?

**Comment by riversand963** [comment]
> @riversand963 Thanks for the explanation about the reasoning behind handling CF add/drop. With this recovery, we don't support cross CF consistency. We may have to revisit later for cases when WAL is used or atomic flush is enabled. For now, should we just not try to recover if either of those is true?

Agree.
In fact, I realize that `TryRecover` from an inconsistent MANIFEST can result in `DB::Open` failure if WAL is enabled and replayed. So let's disable it for now.

File: db/version_edit_handler.h (hunk 12 above)
Comment (riversand963) [reply db/version_edit_handler.h]: Done

File: db/version_edit_handler.h (hunk 11 above)
Comment (riversand963) [reply db/version_edit_handler.h]: Done

File: db/version_edit_handler.cc (hunk 9 above)
Comment (riversand963) [reply db/version_edit_handler.cc]: Done

**Comment by riversand963** [comment]
@anand1976 , could you PTAL? Thanks

File: db/db_impl/db_impl_open.cc [hunk 14]
```diff
@@ -419,7 +425,17 @@ Status DBImpl::Recover(
     }
   }
   assert(db_id_.empty());
-  Status s = versions_->Recover(column_families, read_only, &db_id_);
+  Status s;
+  bool missing_table_file = false;
```
Comment (anand1976) [inline db/db_impl/db_impl_open.cc]: Where is ```missing_table_file``` used?

File: db/db_impl/db_impl_open.cc [hunk 15]
```diff
@@ -499,7 +515,9 @@ Status DBImpl::Recover(
     // attention to it in case we are recovering a database
     // produced by an older version of rocksdb.
     std::vector<std::string> filenames;
-    s = env_->GetChildren(immutable_db_options_.wal_dir, &filenames);
+    if (!immutable_db_options_.best_efforts_recovery) {
```
Comment (anand1976) [inline db/db_impl/db_impl_open.cc]: Is just ignoring WAL files when best_efforts_recovery is enabled correct? Or should the recovery fail if there are non-empty WAL files?

File: db/version_builder.h [hunk 16]
```diff
@@ -44,5 +46,21 @@ class VersionBuilder {
...
+ public:
+  explicit BaseReferencedVersionBuilder(ColumnFamilyData* cfd);
+  BaseReferencedVersionBuilder(ColumnFamilyData* cfd, Version* v);
+  ~BaseReferencedVersionBuilder();
+  VersionBuilder* version_builder() const { return version_builder_.get(); }
+
+ private:
+  std::unique_ptr<VersionBuilder> version_builder_;
+  Version* version_;
+  const bool should_ref_cfd_version_;
```
Comment (anand1976) [inline db/version_builder.h]: I suggest adding a comment to clarify what this is used for.

File: db/version_edit_handler.cc [hunk 17]
```diff
@@ -0,0 +1,591 @@
...
+    if (!s.ok()) {
+      break;
+    }
+    if (edit.has_db_id_) {
+      version_set_->db_id_ = edit.GetDbId();
+      if (db_id != nullptr) {
+        *db_id = version_set_->db_id_;
+      }
+    }
+    s = read_buffer_.AddEdit(&edit);
```
Comment (anand1976) [inline db/version_edit_handler.cc]: We add every version edit to ```read_buffer_```, but only clear it if its an atomic group and the buffer is replayed?

File: db/version_set_test.cc [hunk 18]
```diff
@@ -1279,6 +1331,704 @@ INSTANTIATE_TEST_CASE_P(
...
+    ColumnFamilyData* cfd = versions_->GetColumnFamilySet()->GetColumnFamily(
+        kUnknownColumnFamilyName);
+    ASSERT_EQ(nullptr, cfd);
+  }
+}
+
+INSTANTIATE_TEST_CASE_P(
+    BestEffortRecovery, VersionSetTestEmptyDb,
+    testing::Combine(
+        testing::Bool(), testing::Bool(),
```
Comment (anand1976) [inline db/version_set_test.cc]: Can you add comments on what each parameter is for?

File: include/rocksdb/options.h [hunk 19]
```diff
@@ -1130,6 +1130,16 @@ struct DBOptions {
   //
   // Default: nullptr
   std::shared_ptr<FileChecksumFunc> sst_file_checksum_func = nullptr;
+
+  // By default, RocksDB recovery fails if any table file referenced in
+  // MANIFEST are missing after scanning the MANIFEST.
+  // Best-efforts recovery is another recovery mode that
+  // tries to restore the database to the most recent point in time without
+  // missing file.
+  // Currently not compatible with atomic flush. Furthermoare, WAL files will
```
Comment (anand1976) [inline include/rocksdb/options.h]: Typo - Furthermoare

Comment (riversand963) [reply include/rocksdb/options.h]: Will fix

File: db/db_impl/db_impl_open.cc (hunk 14 above)
Comment (riversand963) [reply db/db_impl/db_impl_open.cc]: It's used in testing.

File: db/version_builder.h (hunk 16 above)
Comment (riversand963) [reply db/version_builder.h]: I am actually rethinking this. Maybe I can try to remove this variable. If it's non-trivial, will add a comment.

File: db/db_impl/db_impl_open.cc (hunk 15 above)
Comment (riversand963) [reply db/db_impl/db_impl_open.cc]: On the one hand, WAL contains data ahead of SSTs (not considering WAL with prepared entries). On the other hand, it is very common to have non-empty WAL even if db closed cleanly.
I think we should not report best-efforts recovery as failure.

File: db/version_set_test.cc (hunk 18 above)
Comment (riversand963) [reply db/version_set_test.cc]: Sure.

File: db/db_impl/db_impl_open.cc (hunk 15 above)
Comment (riversand963) [reply db/db_impl/db_impl_open.cc]: Furthermore, I think we need to bump up cfd->log_number and min_log_number_to_keep after we ignore the existing WALs.

File: db/version_edit_handler.cc (hunk 17 above)
Comment (riversand963) [reply db/version_edit_handler.cc]: As https://github.com/facebook/rocksdb/blob/master/db/version_set.cc#L3513:L3518 shows, AtomicGroupReadBuf does not actually buffer a non-atomic edit.

File: db/version_builder.h (hunk 16 above)
Comment (riversand963) [reply db/version_builder.h]: Removed, thus no need for comment any more.

File: db/db_impl/db_impl_open.cc (hunk 15 above)
Comment (anand1976) [reply db/db_impl/db_impl_open.cc]: I was thinking about cross-CF consistency. Do we implicitly guarantee it if the user uses WAL? We don't guarantee it with the best efforts recovery right? Maybe we can just mention it in READ.md.

Comment (riversand963) [reply db/db_impl/db_impl_open.cc]: Good point. User may very likely assume that cross-CF consistency is guaranteed if they enable WAL. We should call this out in HISTORY.md.
